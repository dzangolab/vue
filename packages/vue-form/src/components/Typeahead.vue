<template>
  <div ref="dzangolabVueFormTypeahead" :class="`field typeahead ${name}`">
    <label v-if="label" :for="name">
      {{ label }}
    </label>
    <Field
      v-slot="{ meta }"
      v-bind="{ modelValue }"
      :name="name"
      :rules="fieldSchema"
    >
      <DebouncedInput
        :id="`input-field-${name}`"
        :class="{
          invalid: meta.touched && !meta.valid,
          valid: meta.dirty && meta.valid && Object.keys(props.schema).length,
        }"
        :model-value="inputValue"
        :debounce-time="debounceTime"
        :disabled="disabled"
        :placeholder="placeholder"
        :type="type"
        tabindex="0"
        @update:model-value="onInput"
      />
      <ErrorMessage v-if="!filteredSuggestions.length" :name="name" />
    </Field>
    <ul v-if="showSuggestions" class="suggestion-list">
      <li
        v-for="suggestion in filteredSuggestions"
        :key="suggestion.label"
        class="suggestion"
        @click="onSelect(suggestion)"
      >
        {{ suggestion.label }}
      </li>
    </ul>
  </div>
</template>

<script lang="ts">
export default {
  name: "Typeahead",
};
</script>

<script setup lang="ts">
import { DebouncedInput } from "@dzangolab/vue3-ui";
import { toFieldValidator } from "@vee-validate/zod";
import { onClickOutside } from "@vueuse/core";
import { ErrorMessage, Field } from "vee-validate";
import { computed, ref } from "vue";
import { z } from "zod";

import type { SelectOption } from "../types";
import type { PropType } from "vue";

const props = defineProps({
  debounceTime: {
    default: 500,
    required: false,
    type: Number,
  },
  disabled: {
    default: false,
    type: Boolean,
  },
  label: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  modelValue: {
    default: "",
    required: false,
    type: [String, Number] as PropType<string | number | null | undefined>,
  },
  name: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  suggestions: {
    required: true,
    type: Array as PropType<SelectOption[]>,
  },
  placeholder: {
    default: "",
    type: String,
  },
  schema: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<z.ZodType<string | number | object>>,
  },
  type: {
    default: "text",
    type: String,
  },
});

const dzangolabVueFormTypeahead = ref(null);

const inputValue = ref(props.modelValue);

const showSuggestions = ref<boolean>(false);

const emit = defineEmits(["update:modelValue"]);

onClickOutside(dzangolabVueFormTypeahead, (event) => {
  showSuggestions.value = false;
});

const fieldSchema = Object.keys(props.schema).length
  ? toFieldValidator(props.schema)
  : null;

const filteredSuggestions = computed(() => {
  return props.suggestions.filter((suggestion) =>
    (suggestion.value as string)
      .toLowerCase()
      .includes((inputValue.value as string).toLowerCase())
  );
});

const onInput = (value: string | number) => {
  inputValue.value = value;
  showSuggestions.value = true;

  emit("update:modelValue", value);
};

const onSelect = (option: SelectOption) => {
  inputValue.value = option.value;
  showSuggestions.value = false;

  emit("update:modelValue", option.value);
};
</script>

<style lang="css">
.typeahead {
  position: relative;
  display: inline-block;
  width: 100%;
}

.typeahead input {
  margin-top: 0.75rem;
}

.typeahead input:focus {
  position: relative;
  z-index: 1000;
}

li.suggestion {
  padding: 10px;
  cursor: pointer;
}

li.suggestion:hover {
  --_typeahead-suggestion-hover-color: var(--field-option-hover-color, #e5e5e5);

  background-color: var(--_typeahead-suggestion-hover-color);
  cursor: pointer;
}

ul.suggestion-list {
  --_typeahead-suggestions-border-radius: var(
    --input-field-border-radius,
    0.5rem
  );

  background-color: #fff;
  border: 1px solid #ccc;
  border-radius: var(--_typeahead-suggestions-border-radius);
  border-top: none;
  list-style-type: none;
  margin: 0;
  max-height: 10rem;
  overflow-y: scroll;
  padding: 0;
  position: absolute;
  width: 100%;
  z-index: 999;
}
</style>
