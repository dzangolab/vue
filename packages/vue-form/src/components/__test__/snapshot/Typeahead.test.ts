import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import Typeahead from "../../Typeahead.vue";

describe("Typeahead", () => {
  const suggestionItems = [
    {
      value: "Roslabel",
      label: "Roslabel - A fragrant flower often associated with romance",
    },
    {
      value: "Tulip",
      label: "Tulip - A bright, cup-shaped flower popular in spring",
    },
    {
      value: "Daisy",
      label: "Daisy - A simple, white flower with a yellow center",
    },
    {
      value: "Sunflower",
      label: "Sunflower - A tall plant with a large, yellow flower head",
    },
    {
      value: "Lily",
      label: "Lily - An elegant flower often used in bouquets",
    },
    {
      value: "Orchid",
      label: "Orchid - An exotic flower with a unique shape",
    },
    {
      value: "Marigold",
      label: "Marigold - A vibrant, orange or yellow flower",
    },
    {
      value: "Lavender",
      label: "Lavender - A fragrant flower known for its calming scent",
    },
    {
      value: "Peony",
      label: "Peony - A lush, full flower often used in weddings",
    },
    {
      value: "Chrysanthemum",
      label: "Chrysanthemum - A hardy flower with a variety of colors",
    },
    {
      value: "Daffodil",
      label: "Daffodil - A bright, yellow flower that blooms in spring",
    },
    {
      value: "Hydrangea",
      label: "Hydrangea - A large, cluster-like flower that changes color",
    },
  ];

  it("matches snapshot", () => {
    const wrapper = mount(Typeahead, {
      global: {
        mocks: {
          showSuggestions: true,
        },
      },
      props: {
        label: "Input",
        placeholder: "Placeholder",
        suggestions: suggestionItems,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
