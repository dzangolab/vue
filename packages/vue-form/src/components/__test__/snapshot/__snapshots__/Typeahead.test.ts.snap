// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typeahead > matches snapshot 1`] = `
<div
  class="field typeahead"
>
  <label
    for=""
  >
    Input
  </label>
  
  <div
    class="debounce-input"
    id="input-field-"
    tabindex="0"
  >
    <input
      placeholder="Placeholder"
      type="text"
    />
  </div>
  <!--v-if-->
  
  <ul
    class="suggestion-list"
  >
    
    <li
      class="suggestion"
    >
      Roslabel - A fragrant flower often associated with romance
    </li>
    <li
      class="suggestion"
    >
      Tulip - A bright, cup-shaped flower popular in spring
    </li>
    <li
      class="suggestion"
    >
      Daisy - A simple, white flower with a yellow center
    </li>
    <li
      class="suggestion"
    >
      Sunflower - A tall plant with a large, yellow flower head
    </li>
    <li
      class="suggestion"
    >
      Lily - An elegant flower often used in bouquets
    </li>
    <li
      class="suggestion"
    >
      Orchid - An exotic flower with a unique shape
    </li>
    <li
      class="suggestion"
    >
      Marigold - A vibrant, orange or yellow flower
    </li>
    <li
      class="suggestion"
    >
      Lavender - A fragrant flower known for its calming scent
    </li>
    <li
      class="suggestion"
    >
      Peony - A lush, full flower often used in weddings
    </li>
    <li
      class="suggestion"
    >
      Chrysanthemum - A hardy flower with a variety of colors
    </li>
    <li
      class="suggestion"
    >
      Daffodil - A bright, yellow flower that blooms in spring
    </li>
    <li
      class="suggestion"
    >
      Hydrangea - A large, cluster-like flower that changes color
    </li>
    
  </ul>
</div>
`;
